lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.cpp

```diff
@@ -0,0 +1,212 @@
+
+    // Extract value at current index
+    auto newExtractOp = builder.create<tensor::ExtractOp>(
+        extractOp.getTensor(), forOp.getInductionVar());
+    setSecretnessOfResults(solver, extractOp->getOperands(),
+                           extractOp->getResults());
+
+    auto ifOp = builder.create<scf::IfOp>(
+        cond,
+        [&](OpBuilder &b, Location loc) {
+          // Yield value extracted at index
+          b.create<scf::YieldOp>(loc, newExtractOp.getResult());
+        },
+        [&](OpBuilder &b, Location loc) {
+          // Yield dummyValue
+          b.create<scf::YieldOp>(loc, forOp.getRegionIterArgs().front());
+        });
+    // TODO (#888): Secretness is valid for this region-bearing op only when
+    // index is strictly secret; secretness won't be valid if index has unknown
+    // secretness (i.e., when transformation occurs in the conservative case)
+    setSecretnessOfResults(solver, ifOp->getOperands(), ifOp->getResults());
+
+    // Create YieldOp for affine.for
+    SmallVector<Value> results(ifOp->getOpResults());
+    builder.create<affine::AffineYieldOp>(results);
+
+    // Replace the old tensor.insert op with forOp's result
+    rewriter.replaceOp(extractOp, forOp);
+
+    return success();
+  }
+
+ private:
+  DataFlowSolver *solver;
+
+  static inline void setValueToSecretness(DataFlowSolver *solver, Value value,
+                                          Secretness secretness) {
+    auto *lattice = solver->getOrCreateState<SecretnessLattice>(value);
+    solver->propagateIfChanged(lattice, lattice->join(secretness));
+  }
+
+  static inline void setSecretnessOfResults(DataFlowSolver *solver,
+                                            OperandRange operands,
+                                            ResultRange results) {
+    auto resultSecretness = Secretness();
+
+    // Handle operations without operands (e.g. arith.constant)
+    if (operands.empty()) {
+      resultSecretness.setSecretness(false);
+    }
+
+    bool isUninitializedOpFound = false;
+
+    for (auto operand : operands) {
+      auto operandSecretness =
+          solver->getOrCreateState<SecretnessLattice>(operand)->getValue();
+
+      if (!operandSecretness.isInitialized()) {
+        // Keep record if operand is uninitialized
+        isUninitializedOpFound = true;
+      }
+      resultSecretness = Secretness::join(resultSecretness, operandSecretness);
+      if (resultSecretness.isInitialized() && resultSecretness.getSecretness())
+        break;
+    }
+
+    if (isUninitializedOpFound) {
+      if (resultSecretness.isInitialized() &&
+          !resultSecretness.getSecretness()) {
+        resultSecretness = Secretness();
+      }
+    }
+
+    for (auto result : results) {
+      auto *lattice = solver->getOrCreateState<SecretnessLattice>(result);
+      solver->propagateIfChanged(lattice, lattice->join(resultSecretness));
+    }
+  }
+};
```

Comment:
I'm not a fan of this, as it duplicates logic that should only be in the secrectness analysis (I guess one could say it violates [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and [SoC](https://en.wikipedia.org/wiki/Separation_of_concerns))
I think you should be able to avoid this, as you only ever create operations for which you KNOW what the secretness should be.